### ISSUE TITLE
Make failing migrations disable dependent features

### ISSUE THREAD

**Issue by Piedone** [OrchardCMS/Orchard#5639]
When a migration step fails it certainly means that the feature is in an uncertain (but erroneous) state. IMO it's safe to assume here that features depending on the feature with the failed migration won't work properly. Thus I suggest that when a migration step fails all features depending on the feature of the migration be disabled.

I'd imagine a FeatureManager.Disable() added to DataMigrationManager.Update()'s catch, with a corresponding log entry. Also maybe a warning on the Features screen: if there are features to be updated (i.e. those that have failed their automatic migration) there  could be warning that all dependent features were automatically disabled.

**Comment by CSurieux** [comment]
And why don't simply do a rollback of the failled migration?

**Comment by bleroy** [comment]
That should already be the case: the transaction should be rolled back in this case.

**Comment by Piedone** [comment]
Yes, it's not about the transaction: it only rolls back the migration step. This issue is about the higher level: turning off depending features (even if they were already turned on a long time ago).

**Comment by bleroy** [comment]
Right, but if the transaction is rolled back, why would you want that?

**Comment by Piedone** [comment]
Because that won't change the state of already enabled features (in the case you add a new, failing migration step to an already enabled feature that has other, already enabled features depending on it). I'm not sure about this but I think a failing migration won't prevent dependent features from being newly enabled either (i.e. if feature B depends on A, you enable B but the migration of A fails, still both features will be enabled).

Keep in mind that (although we could change this, but it has its reasons) migrations run in their own transaction. Enabling a feature that automatically enables its dependencies is as many transactions as the number of migration classes.

**Comment by bleroy** [comment]
I see. Thanks for the explanation. OK then, I agree, but we need to explain very carefully what just happened when that happens, otherwise it could be pretty confusing.

**Comment by CSurieux** [comment]
+1
